crates/typst-library/src/layout/raster.rs

```diff
@@ -0,0 +1,1788 @@
+        match self {
+            Self::Repeated(repeated) => repeated,
+            Self::NotRepeated(not_repeated) => not_repeated,
+        }
+    }
+
+    /// Returns `Some` if the value is repeated, `None` otherwise.
+    pub fn as_repeated(&self) -> Option<&T> {
+        match self {
+            Self::Repeated(repeated) => Some(repeated),
+            Self::NotRepeated(_) => None,
+        }
+    }
+}
+
+/// Indicates whether the line should be drawn before or after the track with
+/// its index. This is mostly only relevant when gutter is used, since, then,
+/// the position after a track is not the same as before the next
+/// non-gutter track.
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum LinePosition {
+    /// The line should be drawn before its track (e.g. hline on top of a row).
+    Before,
+    /// The line should be drawn after its track (e.g. hline below a row).
+    After,
+}
+
+/// A raster entry.
+pub enum Entry<'a> {
+    /// An entry which holds a cell.
+    Cell(Cell<'a>),
+    /// An entry which is merged with another cell.
+    Merged {
+        /// The index of the cell this entry is merged with.
+        parent: usize,
+    },
+}
+
+impl<'a> Entry<'a> {
+    /// Obtains the cell inside this entry, if this is not a merged cell.
+    fn as_cell(&self) -> Option<&Cell<'a>> {
+        match self {
+            Self::Cell(cell) => Some(cell),
+            Self::Merged { .. } => None,
+        }
+    }
+}
+
+/// Any raster child, which can be either a header or an item.
+enum ResolvableRasterChild<T: ResolvableCell, I> {
+    Header { repeat: bool, span: Span, items: I },
+    Footer { repeat: bool, span: Span, items: I },
+    Item(ResolvableRasterItem<T>),
+}
+
+/// A raster of cells, including the columns, rows, and cell data.
+pub struct Raster<'a> {
+    /// The raster cells.
+    pub entries: Vec<Entry<'a>>,
+    /// The column tracks including gutter tracks.
+    pub cols: Vec<Sizing>,
+    /// The row tracks including gutter tracks.
+    pub rows: Vec<Sizing>,
+    /// The vertical lines before each column, or on the end border.
+    /// Gutter columns are not included.
+    /// Contains up to 'cols_without_gutter.len() + 1' vectors of lines.
+    pub vlines: Vec<Vec<Line>>,
+    /// The horizontal lines on top of each row, or on the bottom border.
+    /// Gutter rows are not included.
+    /// Contains up to 'rows_without_gutter.len() + 1' vectors of lines.
+    pub hlines: Vec<Vec<Line>>,
+    /// The repeatable header of this raster.
+    pub header: Option<Repeatable<Header>>,
+    /// The repeatable footer of this raster.
+    pub footer: Option<Repeatable<Footer>>,
+    /// Whether this raster has gutters.
+    pub has_gutter: bool,
+}
+
+/// Position of a cell inside a raster.
+///
+/// Given an index `i` and the number of columns `c`,
+/// the x-coordinate of `i` is `i % c` and
+/// the y-coordinate of `i` is `i / c`.
+type Idx = usize;
+
+#[derive(Default)]
+struct State<'a> {
```

Comment:
These refactors to grid resolving, such as with the introduction of `State`, is not really necessary for your goal, right? If that's the case, could we please move these changes to a separate PR? That'd make it much simpler to review and discuss this PR. Thank you.

Reply:
I agree with you. I have opened a new PR #5585 that tries to focus on only moving code.